src/math/trig.c

```diff
@@ -9,568 +9,95 @@
-	int i1;
-
-	i1 = idx & (2 * SINE_NQUART - 1);
-	if (i1 > SINE_NQUART)
-		i1 = 2 * SINE_NQUART - i1;
-
-	if (idx > 2 * SINE_NQUART)
-		s = -sine_table[i1];
-	else
-		s = sine_table[i1];
-
-	return s;
-}
-
-/* Compute fixed point sine with table lookup and interpolation */
-int32_t sin_fixed(int32_t w)
+int32_t sin_fixed(int32_t th_rad_fxp)
 {
-	int32_t frac;
-	int32_t s0;
-	int32_t s1;
-	int32_t delta;
-	int64_t idx;
-	int64_t sine;
-	int64_t idx_tmp;
+	/*cordic_atan2_lookup_table = atan(2.^-(0:N-1)) N = 31
+	 *CORDIC Gain is cordic_gain = prod(sqrt(1 + 2.^(-2*(0:31-1))))
+	 *Inverse CORDIC Gain,inverse_cordic_gain = 1 / cordic_gain
+	 */
+	static const int32_t cordicsine_lookup[CORDICSINE_TABLE_SIZE] = { 843314857, 497837829,
+	263043837, 133525159, 67021687, 33543516, 16775851, 8388437, 4194283, 2097149,
+	1048576, 524288, 262144, 131072, 65536, 32768, 16384, 8192, 4096, 2048, 1024,
+	512, 256, 128, 64, 32, 16, 8, 4, 2, 1 };
+	int32_t b_idx;
+	int32_t b_yn;
+	int32_t xn;
+	int32_t xtmp;
+	int32_t ytmp;
+	int sign = 1;
+	/* 652032874 , deg = 69.586061*/
+	const int32_t CORDICSINE_LUT_Q29FL  =  Q_CONVERT_FLOAT(1.214505869895220, 29);
+	/* 421657428 , deg = 90.000000 */
+	const int32_t CORDICSINE_PIOVERTWO_Q28FL  = Q_CONVERT_FLOAT(M_PI / 2, 28);
+	/* 843314857,  deg = 90.000000  */
+	const int32_t CORDICSINE_PIOVERTWO_Q29FL  = Q_CONVERT_FLOAT(M_PI / 2, 29);
+	/* 1686629713, deg = 90.000000  */
+	const int32_t CORDICSINE_PIOVERTWO_Q30FL  = Q_CONVERT_FLOAT(M_PI / 2, 30);
 
-	/* Q4.28 x Q12.20 -> Q16.48 */
-	idx_tmp = (int64_t)w * SINE_C_Q20;
-	idx = (idx_tmp >> 48); /* Shift to Q0 */
-	idx_tmp = idx_tmp >> 17; /* Shift to Q16.31 */
-	idx_tmp = idx_tmp - (idx << 31); /* Get fraction */
-	frac = (int32_t)idx_tmp; /* Q1.31 */
-	s0 = sine_lookup(idx); /* Q1.31 */
-	s1 = sine_lookup(idx + 1); /* Q1.31 */
-	delta = s1 - s0; /* Q1.31 */
-	/* All Q1.31 */
-	sine = s0 + q_mults_32x32(frac, delta, Q_SHIFT_BITS_64(31, 31, 31));
+	/*Addition or subtraction by a multiple of pi/2 is done in the data type
+	 *of the input. When the fraction length is 29, then the quantization error
+	 *introduced by the addition or subtraction of pi/2 is done with 29 bits of
+	 *precision.Input range of cordicsin must be in the range [-2*pi, 2*pi),
+	 *a signed type with fractionLength = wordLength-4 will fit this range
+	 *without overflow.Increase of fractionLength makes the addition or
+	 *subtraction of a multiple of pi/2 more precise
+	 */
+	if (th_rad_fxp > CORDICSINE_PIOVERTWO_Q28FL) {
+		if ((th_rad_fxp - CORDICSINE_PIOVERTWO_Q29FL) <= CORDICSINE_PIOVERTWO_Q28FL) {
+			th_rad_fxp -= CORDICSINE_PIOVERTWO_Q29FL;
+			sign  = -1;
+		} else {
+			th_rad_fxp -= CORDICSINE_PIOVERTWO_Q30FL;
+			sign  = 1;
+		}
+	} else if (th_rad_fxp < -CORDICSINE_PIOVERTWO_Q28FL) {
+		if ((th_rad_fxp + CORDICSINE_PIOVERTWO_Q29FL) >= -CORDICSINE_PIOVERTWO_Q28FL) {
+			th_rad_fxp += CORDICSINE_PIOVERTWO_Q29FL;
+			sign  = -1;
+		} else {
+			th_rad_fxp += CORDICSINE_PIOVERTWO_Q30FL;
+			sign  = 1;
```

Comment:
Can this `sign = 1; `set be left out?

Reply:
Yes. this can be done as it is set to 1 in the beginning.